Implement Homebrew as the system deployer for macOS.

simonjwright opened the pull request:
See Issue #1172.

It's not clear to me how we could deal with MacPorts as an alternate deployer, if anyone wants to do the work (won't be me).

I'm not sure about  `System_Package_For_Tool` in `src/alire/alire-utils-tools.adb`; I've treated macOS like all the other distributions so as to get a build, but I don't understand how/when this is used.

simonjwright committed:
Implement Homebrew as the system deployer for macOS.

  * src/alire/alire-origins-deployers-system-homebrew.adb: new.
  * src/alire/alire-origins-deployers-system-homebrew.ads: new.
  * src/alire/alire-origins-deployers-system.adb (Platform_Deployer):
      When the platform distro manager is Homebrew, return Homebrew.Deployer.
  * src/alire/alire-platforms.ads (Distributions): include macOS.
    (Package_Managers): add Homebrew.
    (Distro_Manager): for macOS, use Homebrew.
  * src/alire/alire-utils-tools.adb (System_Package_For_Tool): treat MacOS
      like other distributions.
  * src/alire/os_macos/alire-platforms-current__macos.adb
    (Detected_Distribution): return Platforms.MacOS.
  * testsuite/drivers/helpers.py (on_macos()): new.
    (distribution()): return 'MACOS' if on_macos() is True.

; Squashed commit of the following:

; commit 0110299ceb66f58cd4793fd78a69fc5b100b4ccf
; Author: Simon Wright <simon@pushface.org>
; Date:   Fri Sep 9 17:45:04 2022 +0100

;     Restore test to correct sense.

; commit d396b6b54a73bbbfd59071539bf6268bccce3cf0
; Author: Simon Wright <simon@pushface.org>
; Date:   Fri Sep 9 17:34:54 2022 +0100

;     Replace assertions in Homebrew deployer with exceptions

; commit d7603110b9543bdaac9516fa844599e87fac70aa
; Author: Simon Wright <simon@pushface.org>
; Date:   Mon Sep 5 14:14:03 2022 +0100

;     Recognise macOS as a distribution.

;       * testsuite/drivers/helpers.py (on_macos()): new.
;         (distribution()): check on_macos(). Return 'MACOS' if true.

; commit 8703acfeb8fdf336d862cefc16706eac5d3f0758
; Author: Simon Wright <simon@pushface.org>
; Date:   Sun Sep 4 17:33:13 2022 +0100

;     Use 'brew info' to get available, installed versions.

; commit 8b22b904e19d64fecb89233e6527c2faf655d982
; Author: Simon Wright <simon@pushface.org>
; Date:   Wed Aug 31 15:54:51 2022 +0100

;     Continuing.

; commit a9821ed9a605e30b5699a49df6fffd11224ab931
; Author: Simon Wright <simon@pushface.org>
; Date:   Mon Aug 22 17:09:48 2022 +0100

;     First changes for macOS Homebrew.

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,273 @@
+   use GNATCOLL.JSON;
+
+   Homebrew_Prefix : constant String
+     := Ada.Environment_Variables.Value ("HOMEBREW_PREFIX", "");
```
Across the project we use `Alire.OS_Lib.Getenv` to get environment variable . We should probably use `Ada.Environment_Variables.Value` but for consistency let's use keep `Alire.OS_Lib.Getenv` everywhere.

simonjwright replied:
OK (but may become OBE depending on how we resolve the Homebrew/MacPorts issue)

simonjwright replied:
Actually, I suspect that this code will only get called at all if the `Distribution` is `Homebrew`? in which case the question is moot (but I would have to change the implementation of `Already_Installed `to use the `Installed_Version` found by `Get_Info`, because my lazy implementation relies on knowing `Homebrew_Prefix`).

And the tests for `Homebrew_Present` would be redundant.

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,273 @@
+         Map_JSON_Object (Get (Arr, 1), Info_Callback'Access);
+      end;
+
+   --  exception
```
Dead code?

simonjwright replied:
Dead comment, anyway! OK.

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,273 @@
+   use GNATCOLL.JSON;
+
+   Homebrew_Prefix : constant String
+     := Ada.Environment_Variables.Value ("HOMEBREW_PREFIX", "");
```
Is the `HOMEBREW_PREFIX` env var something common for Homebrew users? Or a requirement only for Alire users?

simonjwright replied:
The 'normal' way for users to set up to use Homebrew is to do something like
```
eval "$(/where/ever/bin/brew shellenv)"
``` 
in their `.bash_profile` or equivalent, where `brew shellenv` returns (on my aarch64 Mac)
```
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```
Of course they could just set the `PATH` themselves. Which means no `HOMEBREW_PREFIX` etc; but then consider [this](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/options.gpr#L36) and [this](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/options.gpr#L63) in my tash crate (to be announced), 

The less-obvious issue is that Homebrew on `x86_64-apple-darwin` installs in `/usr/local` by default, but on `aarch64` it installs in `/opt/homebrew` (so as not to confuse people who have both architectures installed???) -- the `/usr/local` thing was my chief reason for rejecting Homebrew; you can change it, e.g. on my x86_64 Mac I have it in `~/homebrew`, which has the disadvantage that packages have to be rebuilt from source rather than being available in binary.

simonjwright replied:
The bottom line is, we can probably get away without setting up these environment variables for `alr` itself (check whether `brew` is on the `PATH`), but a crate that depends on an external will need them.

Fabien-Chouteau replied:
If `eval "$(/where/ever/bin/brew shellenv)"` is the normal way then using `HOMEBREW_PREFIX` sounds like a perfectly good solution to me :+1:

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,273 @@
+
+   begin
+      Trace.Debug ("detect? " & This.Base.Package_Name);
+      if Homebrew_Present then
```
Detection of the "distrib" should be done in the platform specific code. There's no need to do it here as far as I know.

simonjwright replied:
OK

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,273 @@
+
+      function Get_Version_From_String (Candidate : String)
+                                       return Version_Outcomes.Outcome;
+      function Get_Version_From_String (Candidate : String)
```
The `Semantic_Versioning` library is already providing this kind of conversion. Is there a reason not to use it?

simonjwright replied:
No reason. The existing code will (try to) detect format errors, is it OK to let `Semantic_Versioning` raise an exception in this case? would said exception need to be handled? (it doesn't seem to be handled in other uses).

My code would report libtcl's version as 8.6.12, Semantic_Versioning reports 8.6.12+_1 - would that be a problem? (building my test code with this works fine).

Fabien-Chouteau replied:
@mosteo I think we need your input here.

mosteo replied:
The extra build part should not be a problem. We have to be careful if distributions use `-` in a sense different than semantic versioning pre-releases, though, to avoid mis-reporting versions.

These detectors shouldn't raise; in the worst case we prefer the external to be reported as undetected. But then, relaxed semantic versioning will accept anything that starts with a number as a version.

Fabien-Chouteau commented on src/alire/alire-platforms.ads:
```diff
@@ -40,6 +40,7 @@ package Alire.Platforms with Preelaborate is
                           Rhel, -- RedHat Enterprise Linux
                           Centos,
                           Fedora,
+                          MacOS,
```
This entry should be `Homebew` and not `MacOS`, for the same reasons that we have `Msys2` and not `Windows`:
 - You can use Alire on MacOS without Homebew
 - You can have other system package managers on MacOS (Macports)

simonjwright replied:
OK

Joebeazelman replied:
Homebrew also supports Windows 10 subsystem and Linux. You might need to create a package matrix or perhaps another environment variable for the parent package manager, i.e. how was Alire installed regardless of platform. This might be a consideration if, for instance, some operation might need the information.

simonjwright replied:
> Homebrew also supports Windows 10 subsystem and Linux. You might need to create a package matrix or perhaps another environment variable for the parent package manager, i.e. how was Alire installed regardless of platform. This might be a consideration if, for instance, some operation might need the information.

When you say "You might need to create ..." I hope you don't mean me? This sounds way outside my comfort zone

mosteo replied:
> When you say "You might need to create ..." I hope you don't mean me? This sounds way outside my comfort zone

I guess it was said as hypothetical future feature. In any case, Alire currently only supports one package manager per distro. But, certainly, one can use combos [os x distro] in the externals, although the need has not yet arisen.

Which makes me think that, as we have `macOS` as the OS, maybe it's more appropriate to use `homebrew` as the distro just like `msys2` is used on Windows. This would enable using other package manager on macOS if someone wants to come up with it.

simonjwright replied:
Yes, that’s how I’m approaching it. I’m tidying up the update to this PR.

Fabien-Chouteau commented on src/alire/os_macos/alire-platforms-current__macos.adb:
```diff
@@ -9,7 +9,7 @@ package body Alire.Platforms.Current is
 
    function Detected_Distribution return Platforms.Distributions is
-      (Platforms.Distro_Unknown);
+      (Platforms.MacOS);
```
This is where you detect the `HOMEBREW_PREFIX` env variable and return `Platforms.Homebew` if present, `Platforms.Distro_Unknown` otherwise.

simonjwright replied:
OK

Fabien-Chouteau commented on testsuite/drivers/helpers.py:
```diff
@@ -89,6 +93,8 @@ def distribution():
 
         return 'DISTRO_UNKNOWN'
 
+    elif on_macos():
```
This will have to be changed according to the comments about MacOS vs Homebew distrib.

simonjwright replied:
This is for CI, and the CI runners have Homebrew installed (but not the environment variables! see [here](https://github.com/simonjwright/tcladashell/blob/8510f86420d0b33bca68441592efc427a7509e5d/.github/workflows/ci-macos.yml#L42)).

So can I punt the Macports issue for the moment and just return `'HOMEBREW'`?

Fabien-Chouteau replied:
I think what we might want to do is have a CI run that use Homebrew (`eval $(brew shellenv)`) and one that doesn't. @mosteo  ?

mosteo replied:
Sounds right to me, just as we test with no package manager on linux and with the supported ones too.

simonjwright replied:
I could certainly set this up.

I did think about checking whether `brew` is on the `PATH`, but users might want to use a different package manager, & if they had both this could lead to confusion; checking for `HOMEBREW_PREFIX` would mean they were OK with it.

Would have to change `testsuite/drivers/helpers.py` to make the same choice: e.g. like [this](https://github.com/simonjwright/alire/commit/50f4fe9ed5cd6aa037ea9a6558a984bde9bb2b46) and [this](https://github.com/simonjwright/alire/commit/913e5573e3d47557d2e750fe2e6e71f2202a3eec), successful run [here](https://github.com/simonjwright/alire/actions/runs/3077781843)

Fabien-Chouteau reviewed (changes requested):
Awesome @simonjwright !

I have a few comments, the main one being about the `MacOS` vs `Homebrew` distrib.

Thanks for your contribution,

simonjwright committed:
Implement Homebrew for MacOS: response to comments on PR#1185

  * .github/workflows/ci-macos.yml (Run test script): renamed to
      "Run test script (without Homebrew)".
    (Run test script (with Homebrew)): added 'eval $(brew shellenv)' before
      running the test script, so that HOMEBREW_PREFIX gets defined.
  * src/alire/alire-origins-deployers-system-homebrew.adb
    (Homebrew_Prefix, Homebrew_Present): removed.
    (Get_Info): named the declare block. Removed commented-out exception
      handler.
    (Already_Installed): use Get_Info's returned Installed_Version's length
      to determine the result.
    (Detect): Get_Version_From_String uses Semantic_Versioning.Parse.
        Don't check for Homebrew_Present, it has to be or we wouldn't be
        called.
  * src/alire/alire-platforms.ads (Distributions): rename MacOS to Homebrew.
    (Distro_Manager): likewise.
  * src/alire/alire-utils-tools.adb (System_Package_For_Tool): MacOS (in
      'case Distribution)' changed to Homebrew.
  * src/alire/os_macos/alire-platforms-current__macos.adb
    (Homebrew_Prefix): new, gets the environment variable "HOMEBREW_PREFIX".
    (Homebrew_Present): new, true if the environment variable is present.
    (Detected_Distribution): checks Homebrew_Present, returns Homebrew
      or Distro_Unknown accordingly.
    (Distribution_Root): returns Homebrew_Prefix if present, otherwise "/".
  * testsuite/drivers/helpers.py (distribution()): if on macOS, return
      HOMEBREW if HOMEBREW_PREFIX is found, DISTRO_UNKNOWN otherwise.

simonjwright committed:
Merge pull request #2 from simonjwright/homebrew-merged-candidate

Implement Homebrew for MacOS: response to comments on PR#1185

mosteo commented:
Looks good to me, outstanding effort, Simon. Since this PR has had quite some discussion, I'm going to wait for @Fabien-Chouteau feedback before merging.

serock committed:
Merging with openSUSE changes in #1173.

  * src/alire/alire-origins-deployers-system.adb: added Homebrew option.
  * src/alire/alire-platforms.ads: likewise.
  * src/alire/alire-utils-tools.adb: likewise.

Fabien-Chouteau commented on src/alire/alire-origins-deployers-system-homebrew.adb:
```diff
@@ -0,0 +1,219 @@
+
+package body Alire.Origins.Deployers.System.Homebrew is
+
+   pragma Warnings (Off);
```
All warnings off for the entire file? You can target a specific warning, but disabling all of them is not ok.

simonjwright committed:
Remove improper pragma in response to comment on #1185

  * src/alire/alire-origins-deployers-system-homebrew.adb: removed
      the innefective pragma Warnings (Off).

Fabien-Chouteau commented:
Thanks @simonjwright, can you fix the conflicts with the OpenSUSE support that was just added? Sorry for that

simonjwright commented:
I did fix the conflicts, and hoped to see CI runs; maybe the reason why they didn’t happen is that Github thinks there are still conflicts?

Fabien-Chouteau commented:
GitHub says there's still one conflic:
```
            System.Zypper.Deployer'(Deployers.Deployer'(Base => From)
<<<<<<< homebrew-merged
                                    with others => <>),
         when Platforms.Homebrew =>
            System.Homebrew.Deployer'(Deployers.Deployer'(Base => From)
                                      with others => <>));
=======
                                    with others => <>));
>>>>>>> master
```

simonjwright commented:
Actually, Github says there are 3 conflicts, but they are in fact the *changes* that I want to make. I don’t see how I can fix this ...

Oh. Perhaps the problem is that I cherry-picked the openSuse commit from master? ... but that introduced conflicts which I fixed, and I don’t see any other trace of it than the commit (cb3ca3b) author ...

Oh². I can make it happen here. Baffled for the time being

mosteo commented:
Let me take a stab at this

actions-user committed:
Merge remote-tracking branch 'alire/master' into homebrew-merged

mosteo commented:
Sorry for stepping in in your fixing, Simon. Merged now!

simonjwright commented:
> Sorry for stepping in in your fixing, Simon. Merged now!

Brilliant! Thanks ..
